### PR TITLE
Allow SchoolGroup users to sign in

### PR DESCRIPTION
**Pull request by cpjmcquillan**
This PR allows SchoolGroup users to sign in with DfE sign in, as well as the AuthenticationFallback. Unless the SchoolGroupJobsFeature is enabled, this will have no effect on the application. If the feature is enabled, a SchoolGroup user can sign in and is redirected to a placeholder page - but no other functionality will work yet.

This PR is part of the work required to complete #1733.

## Changes in this PR
- Allow SchoolGroup users to sign in with DfE sign in
- Allow SchoolGroup users to sign in with Authentication fallback

## Next steps
- Incrementally remove the dependency on vacancies and users being associated to a School, so that the feature can be enabled

### CODE REVIEW

File: app/controllers/hiring_staff/base_controller.rb [hunk 1]
```diff
@@ -24,10 +22,6 @@ def check_terms_and_conditions
     redirect_to terms_and_conditions_path unless current_user&.accepted_terms_and_conditions?
   end
 
-  def current_school
```
Comment (david-mears-dfe) [inline app/controllers/hiring_staff/base_controller.rb:27]: Good catch

File: app/controllers/hiring_staff/sign_in/email/sessions_controller.rb [hunk 2]
```diff
@@ -35,12 +35,14 @@ def change_organisation
...
-    # This flow needs to be adapted when we have implemented school groups/trusts/LAs
-    # It is likely that we will not sign in Trust users as a school.
     information = GetInformationFromLoginKey.new(get_key)
-    @reason_for_failing_sign_in, @schools = information.reason_for_failing_sign_in, information.schools
+    @reason_for_failing_sign_in = information.reason_for_failing_sign_in
+    @schools, @school_groups = information.schools, information.school_groups
     update_session_without_urn(information.details_to_update_in_session)
-    redirect_to auth_email_create_session_path(urn: @schools.first.urn) if @schools&.one?
+    # .to_i used to convert nil to 0
+    only_one_organisation = @schools&.count.to_i + @school_groups&.count.to_i == 1
```
Comment (david-mears-dfe) [inline app/controllers/hiring_staff/sign_in/email/sessions_controller.rb:43]: A minor thing, but I like the Ruby convention where names for boolean variables end with a `?` - makes it easy to see what it's for.

File: app/controllers/hiring_staff/sign_in/email/sessions_controller.rb [hunk 3]
```diff
@@ -54,6 +56,10 @@ def update_session_without_urn(options)
     Rails.logger.warn("Hiring staff signed in via fallback authentication: #{options[:oid]}")
```
Comment (david-mears-dfe) [inline app/controllers/hiring_staff/sign_in/email/sessions_controller.rb:56]: While we're adding UIDs to this controller, let's rename `update_session_without_urn` to e.g. `update_session_without_urn_and_uid` / `without_org_id`

**Comment by david-mears-dfe** [comment]
Can you make sure that your temporary placeholder route is blocked by the feature flag, as an extra layer of certainty? We might instantiate SchoolGroup objects (for some reason) before the feature is switched on. And it's possible that we will mistakenly associate users with a Trust on DSI before the feature is switched on.

```ruby
# console
> FactoryBot.create(:school_group, uid: 4505)
> ENV['FEATURE_SCHOOL_GROUP_JOBS']
=> "false"
```

Screenshot 2020-07-07 at 10 45 02

Screenshot 2020-07-07 at 10 48 50

File: app/controllers/concerns/authentication_concerns.rb [hunk 4]
```diff
@@ -3,14 +3,26 @@ module AuthenticationConcerns
...
+  end
+
   def current_school
-    School.find_by!(urn: session[:urn])
+    @current_school ||= School.find_by!(urn: session[:urn]) if session[:urn].present?
+  end
+
+  def current_school_group
+    if SchoolGroupJobsFeature.enabled?
+      @current_school_group ||= SchoolGroup.find_by!(uid: session[:uid]) if session[:uid].present?
```
Comment (david-mears-dfe) [inline app/controllers/concerns/authentication_concerns.rb:25]: Is it possible for these instance variables to outlive the deletion of the session?

Comment (cpjmcquillan) [reply app/controllers/concerns/authentication_concerns.rb:25]: If they could, that would be no different to what is currently the case. But if the session is deleted then you're not signed in

**Comment by cpjmcquillan** [comment]
> Can you make sure that your temporary placeholder route is blocked by the feature flag, as an extra layer of certainty? We might instantiate SchoolGroup objects before the feature is switched on. And it's possible that we will mistakenly associate users with a Trust on DSI before the feature is switched on. If the latter happens but not the former, then they will be directed to a `Page not found` when they log in with a UID.
> 
> ```ruby
> # console
> > FactoryBot.create(:school_group, uid: 4505)
> # the feature flag is false
> > ENV['FEATURE_SCHOOL_GROUP_JOBS']
> => "false"
> ```
> 
> Screenshot 2020-07-07 at 10 45 02
> 
> Screenshot 2020-07-07 at 10 48 50

You can't sign in with a SchoolGroup now unless the feature is enabled, so I don't think it's necessary to protect a placeholder route that has no real purpose and won't exist when the feature is enabled.

**Comment by david-mears-dfe** [comment]
> You can't sign in with a SchoolGroup now unless the feature is enabled, so I don't think it's necessary to protect a placeholder route that has no real purpose and won't exist when the feature is enabled.

Fair enough.

File: app/controllers/hiring_staff/sign_in/email/sessions_controller.rb [hunk 5]
```diff
@@ -35,17 +35,17 @@ def change_organisation
...
 
   def choose_organisation
-    # This flow needs to be adapted when we have implemented school groups/trusts/LAs
-    # It is likely that we will not sign in Trust users as a school.
     information = GetInformationFromLoginKey.new(get_key)
-    @reason_for_failing_sign_in, @schools = information.reason_for_failing_sign_in, information.schools
-    update_session_without_urn(information.details_to_update_in_session)
-    redirect_to auth_email_create_session_path(urn: @schools.first.urn) if @schools&.one?
+    @reason_for_failing_sign_in = information.reason_for_failing_sign_in
+    @schools, @school_groups = information.schools, information.school_groups
```
Comment (cesidio) [inline app/controllers/hiring_staff/sign_in/email/sessions_controller.rb:40]: minor thing, better to have one assignment per line

**Pull request merged**